src/TestFramework/AbstractTestFrameworkAdapter.php

```diff
@@ -123,34 +124,19 @@ public function getMutantCommandLine(
         );
     }
 
+    /**
+     * @throws UnsupportedTestFrameworkVersion
+     */
     public function getVersion(): string
     {
-        if ($this->version !== null) {
-            return $this->version;
-        }
-
-        $testFrameworkVersionExecutable = $this->commandLineBuilder->build(
-            $this->testFrameworkExecutable,
-            [],
-            ['--version']
-        );
-
-        $process = new Process($testFrameworkVersionExecutable);
-        $process->mustRun();
-
-        $version = 'unknown';
-
-        try {
-            $version = $this->versionParser->parse($process->getOutput());
-        } catch (InvalidArgumentException $e) {
-            $version = 'unknown';
-        } finally {
-            $this->version = $version;
-        }
-
-        return $this->version;
+        return $this->version ?? $this->version = $this->retrieveVersion();
     }
 
+    /**
+     * @throws UnsupportedTestFrameworkVersion
+     */
+    abstract public function checkVersion(): void;
```

Comment:
I would not add it to the main `TestFrameworkAdapter` interface, but instead create a new one, like `MemoryUsageAware` for example. Not sure all test frameworks need it.

Reply:
I feel they should: maybe the version we support is very low and that's fine, but I would really prefer us to throw with a friendly error if the adapter is using an unsupported version rather than letting the user in the dark with a potentially very weird error

Reply:
do you have an idea what the min version for PhpSpec?

Reply:
by the way, Codeception's min version is 3.1.1, and this is controlled by composer itself, so you won't even be able to install it:

https://github.com/maks-rafalko/infection-codeception-adapter/blob/9321a43f57f4ef9ac980bb4cd74376fe900e59a5/composer.json#L56-L58

Reply:
>do you have an idea what the min version for PhpSpec?

No not really :/

>by the way, Codeception's min version is 3.1.1, and this is controlled by composer itself, so you won't even be able to install it:

Not sure to understand what you mean there: the issue is with the PHAR were we instrument a tool (PHPUnit, Codeception & co.) we have no control over

Reply:
> Not sure to understand what you mean there

I mean that *in addition* to this check, we should add the same `conflict` setting to PhpSpec's adapter to fail as earlier as possible for composer-based install. PHAR distribution will be handled by this new `checkVersion` runtime method.

But again, since you are not sure (and me as well) what is the minimum PhpSpec version supported by Infection, and we don't even know criteria, I still think it shouldn't be on the `TestFrameworkAdapter` level, but as a separate interface

Reply:
Good idea 👍